validator_worker/src/sentry_interface.rs

```diff
@@ -1,143 +1,264 @@
+        config: &Config,
+        logging: bool,
+    ) -> Result<Self, ValidatorWorkerError> {
         let whoami = adapter.whoami();
-        let validator = channel.spec.validators.find(|v| v.id === whoami);
-        // assert!(vaildator, 'we can not find validator entry for whoami')
-        let sentry_url = format!("{}/channel/{}", validator.url, channel.id);
-
-        Self {
-            sentry_url,
-            client: Client::new(),
-            logging,
-            channel: channel.to_owned()
+        let client = Client::builder()
+            .timeout(Duration::from_secs(config.fetch_timeout.into()))
+            .build()
+            .unwrap();
+
+        // validate that we are to validate the channel
+        match channel
+            .spec
+            .validators
+            .into_iter()
+            .find(|&v| v.id == whoami)
+        {
+            Some(v) => {
+                let validator_url = format!("{}/channel/{}", v.url, channel.id);
+
+                Ok(Self {
+                    adapter,
+                    validator_url,
+                    client,
+                    logging,
+                    channel: channel.to_owned(),
+                    config: config.to_owned(),
+                })
+            }
+            None => Err(ValidatorWorkerError::InvalidValidatorEntry(
+                "we can not find validator entry for whoami".to_string(),
+            )),
         }
     }
 
-    pub fn propagate() {
+    pub fn propagate(&self, messages: Vec<MessageTypes>) {
+        let serialised_messages: Vec<String> = messages
+            .into_iter()
+            .map(|message| match message {
+                MessageTypes::NewState(new_state) => serde_json::to_string(&new_state).unwrap(),
+                MessageTypes::ApproveState(approve_state) => {
+                    serde_json::to_string(&approve_state).unwrap()
+                }
+                MessageTypes::Heartbeat(heartbeat) => serde_json::to_string(&heartbeat).unwrap(),
+                MessageTypes::RejectState(reject_state) => {
+                    serde_json::to_string(&reject_state).unwrap()
+                }
+                MessageTypes::Accounting(accounting) => serde_json::to_string(&accounting).unwrap(),
+            })
+            .collect();
 
+        for validator in self.channel.spec.validators.into_iter() {
+            let auth_token = self.adapter.get_auth(&validator).unwrap();
+            match propagate_to(
+                &auth_token,
+                self.config.propagation_timeout,
+                &validator,
+                &serialised_messages,
+            ) {
+                Ok(_) => return,
+                Err(e) => handle_http_error(e, &validator.url),
+            }
+        }
     }
 
-    async pub fn get_latest_msg(from: &str, type: &str) -> impl Future<Output = Result<Vec<Channel>, Error>>  {
+    pub async fn get_latest_msg(
+        &self,
+        from: String,
+        message_type: String,
```

Comment:
This won't work with the Enum, that's why I had the complicated code for the `Message` enum in pre-refactor code.

Reply:
hmnnn ... do we really need an enum? it could be just string constants

Reply:
You could, but then you need to match the string and the signature is not verbose enough ( you need to specify in the documentation that this string should be one of the values in those constants or those string literals).
Plus if you forget to add the match statement for a new `MessageType`, and just a note that this can be 1 of many places that we use a `String` for the type, in all places you'll need to match the string and handle all cases

Reply:
I was trying to figure out the same thing:
https://github.com/AdExNetwork/adex-validator-stack-rust/blob/master/domain/src/validator/message.rs#L45

And it got ugly, but I couldn't find anything in serde that will help with this, maybe I've missed something.

Reply:
there's no solution other than to build two separate enums... or use a String

Reply:
I am throwing it just as an option I found:
https://crates.io/crates/enum-kinds
But indeed it's ugly and I might check in serde(i.e. ask) for feature like that or if they will help me to do it in my free time, until then String I guess..